discovery_engine_core/web-api2/src/error/application.rs

```diff
@@ -0,0 +1,258 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+use std::{
+    marker::PhantomData,
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+use actix_web::{
+    body::BoxBody,
+    http::StatusCode,
+    web::ReqData,
+    FromRequest,
+    Handler,
+    HttpResponse,
+    ResponseError,
+};
+use derive_more::{Deref, Display};
+use futures_util::Future;
+use pin_project_lite::pin_project;
+use serde::Serialize;
+use serde_json::Value;
+use thiserror::Error;
+use tracing::error;
+
+use crate::middleware::tracing::RequestId;
+
+use super::json_error::JsonErrorResponseBuilder;
+
+#[derive(Display, Debug, Deref)]
+#[display(fmt = "{}", error)]
+pub struct Error {
+    #[deref(forward)]
+    error: Box<dyn ApplicationError>,
+    request_id: RequestId,
+}
+
+impl<T> From<T> for Error
+where
+    T: ApplicationError,
+{
+    fn from(error: T) -> Self {
+        Self {
+            error: Box::new(error),
+            request_id: RequestId::missing(),
+        }
+    }
+}
+
+impl ResponseError for Error {
+    fn status_code(&self) -> StatusCode {
+        self.error.status_code()
+    }
+
+    fn error_response(&self) -> HttpResponse<BoxBody> {
+        JsonErrorResponseBuilder::render(
+            self.error.kind(),
+            self.request_id,
+            &self.error.encode_details(),
+        )
+        .into_response(self.error.status_code())
+    }
+}
+
+pub trait ApplicationError: std::error::Error + Send + Sync + 'static {
+    fn status_code(&self) -> StatusCode;
+    fn kind(&self) -> &str;
+    fn encode_details(&self) -> Value {
+        Value::Null
+    }
+}
+
+/// Derives [`ApplicationError`] for given type using given http status code.
+macro_rules! derive_application_error {
+    ($name:ident => $code:ident) => {
+        impl ApplicationError for $name {
+            fn status_code(&self) -> StatusCode {
+                StatusCode::$code
+            }
+
+            fn kind(&self) -> &str {
+                stringify!($name)
+            }
+
+            fn encode_details(&self) -> Value {
+                serde_json::to_value(self)
+                    .unwrap_or_else(|err| {
+                        error!(%err, "serializing error details failed");
+                        Value::Null
+                    })
+            }
+        }
+    };
+}
+
+#[derive(Debug, Display, Error, Serialize)]
+/// Given functionality is not implemented.
+pub(crate) struct Unimplemented {
+    pub(crate) functionality: &'static str,
+}
+
+derive_application_error!(Unimplemented => INTERNAL_SERVER_ERROR);
+
+/// Allows to argument errors with an request id by wrapping the endpoint handler.
+///
+/// # Explanation/Why
+///
+/// This is a bit complex as it can only work when wrapping the endpoint handler
```

Comment:
I thought the whole point for moving to actix was to make this less complex then in warp, maybe we should reconsider, or take a closer look at axum, wdyt? Axum is built on top of tower, and looking at it's popularity couldn't we assume it's mature enough?

Reply:
I'm not sure if it covers our needs, but I found a small crate for request id for tower/axum and it looks simple in both implementation and application https://github.com/imbolc/tower-request-id

Reply:
> I'm not sure if it covers our needs,

There are a few docent crates which do add a request id the the header or similar, that is trivial.
E.g. the opentelemtry actix integration does that, too.

The tricky part is to put it into the error response in a ergonomic manner.

> I thought the whole point for moving to actix was to make this less complex then in warp

Most of the parts in this PR are thinks we probably also should do in warp, too. But which where not included in the previous warp specific PR.

So if to compare that PR you would need to compare it with the `WithRequestIdExt` extension trait only. Not the json error, wrapping, error types etc.

I also thought for a while I could just add a middle ware in actix-web, but that failed at runtime because of how axtic-web converts errors into responses *before* it is passed to any middleware even through the middleware accepts an `Result` :face_exhaling: (It's only for errors form other middle-ware and similar but *not* for e.g. routing errors, endpoint returning an error, failing to parse a request parameter or anything like that.)

I personally also prefer axum and changing this PR into a axum one is way less work then creating this PR (most code stays the same, except the startup & routing code and the extension trait is gone, too)

On reason against axum is that it's still rather unstable and we can expect a lot of small changes over time. Especially when GAT is relased there is a good chance not only for axum to majorly change, but also for tower... but probably that is also true for actix-web which probably would go and bump a v5.0 release so idk. :shrug:

Reply:
>The tricky part is to put it into the error response in a ergonomic manner.

Ok, I think I get it now. But it seems that we could achieve such thing in a more readable/easy to understand way, by abstracting framework (wrap or actix) route handlers from the handlers that perform actual business logic, so that our handlers take needed inputs and return a `Result`, and the route handlers are responsible for translating inputs and mapping returned `Result`s for/from functions they are wrapping, smth like:

```rust
pub(crate) async fn user_interactions_route(
    request_id: RequestId,
    user_id: UserId,
    body: InteractionRequestBody,
    state: Arc<AppState>,
) -> Result<impl warp::Reply, Rejection> {
    let _ = handle_user_interactions(user_id, state)
        .await
       // error mapping with req_id
        .map_err(|_| warp::reject::custom(ExternalError(request_id)))?;
    Ok(StatusCode::NO_CONTENT)
}

pub(crate) async fn handle_user_interactions(
    user_id: UserId,
    state: Arc<AppState>,
) -> Result<(), CustomInternalError> {
    // implementation
}
```

Reply:
But that needs to be done with every single endpoint in just the right way meaning it is not just very verbose but also prone to have bugs and just generally bad style.

So you would write a macro to make sure you don't have change bugs due to code repetition.

And at that point you have exactly what I had in the warp PR.

The benefit of the extension trait is it's code you only write once and then never touch and then reuse and the type system makes sure all complex parts are correct.

Reply:
For error mapping we could have one function to reuse everywhere and only repeat the `RequestId` extractor in handlers, it might be more verbose but also simpler to understand and maintain for others, like me ;) That said we can ofc go forward with your solution, but then it would be good if someone more experienced reviewed the code.

Reply:
Yes, but you still have to e.g. keep the parameter list in sync or else you might for example fail an endpoint at runtime because you only removed an extractor in the actual implementation but not the wrapper.

Reply:
@janpetschexain can you add your opinion?

Reply:
i'm tending more towards the midleware solution. even though it's more complex to read, it's easier extendable and less repetetive. i fully get the argument for easier readability of more verbose code though. maybe we can have a sketch/short documentation of the middlewares and endpoints somewhere at the root of the crate (or in another fitting place) to make it easier for new reviewers to follow the code?

Reply:
Be aware that the middle wares do two things:

- wrap errors not from us  (e.g. serde)

- put a request Id in the context when setting up the log span

The code which injects the request id into our errors is just kinda a middleware:

```rust
web::resource("personalized_documents")
    .route(web::get().to(personalized_documents.error_with_request_id()))
```

So you could say it's a middleware around a specific handler insted one around many handlers/services, 
except that actix-web doesn't have something like `.layer()` or similar on handlers, so I had to put in a bit
of type magic to make it work.